### ISSUE TITLE
[srs-4l] Real TETRIO physics

### ISSUE THREAD

**Issue by wirelyre** [wirelyre/tetra-tools#4]
Currently implemented as SRS with six 180º kicks. But TETRIO uses SRS+, a variant that changes I quarter kicks.

All pieces other than I match the current implementation. The correct full kick table for I under TETRIO:

```json
{
    "ne": [[ 2, -2], [ 3, -2], [ 0, -2], [ 0, -3], [ 3,  0]],
    "es": [[-2,  1], [-3,  1], [ 0,  1], [-3,  3], [ 0,  0]],
    "sw": [[ 1, -1], [ 3, -1], [ 0, -1], [ 3,  0], [ 0, -3]],
    "wn": [[-1,  2], [ 0,  2], [-3,  2], [ 0,  0], [-3,  3]],

    "ns": [[ 0, -1], [ 0,  0], [ 1,  0], [-1,  0], [ 1, -1], [-1, -1]],
    "sn": [[ 0,  1], [ 0,  0], [-1,  0], [ 1,  0], [-1,  1], [ 1,  1]],
    "ew": [[-1,  0], [ 0,  0], [ 0,  2], [ 0,  1], [-1,  2], [-1,  1]],
    "we": [[ 1,  0], [ 0,  0], [ 0,  2], [ 0,  1], [ 1,  2], [ 1,  1]]

    "nw": [[ 1, -2], [ 0, -2], [ 3, -2], [ 3, -3], [ 0,  0]],
    "en": [[-2,  2], [-3,  2], [ 0,  2], [-3,  0], [ 0,  3]],
    "se": [[ 2, -1], [ 0, -1], [ 3, -1], [ 0,  0], [ 3, -3]],
    "ws": [[-1,  1], [ 0,  1], [-3,  1], [ 0,  3], [-3,  0]],
}
```

It's not possible to just paste these in, because these reversed rotations don't have [inverted offsets](https://github.com/wirelyre/tetra-tools/blob/a906c589be0474e6edc2e4dc3e26a6d1920ee7ac/srs-4l/src/vector.rs#L772-L796).

Options:
1. `KickTable<const QUARTER: usize, const HALF: usize>`, full kick tables for each physics–shape pair
2. `(&[Kick], &[Kick], &[Kick])`, each component selected at runtime and runtime-sized
3. Special case for I in placement machine

I have low hopes for (2) because I suspect the compiler unrolls the kick loops and it's fast. (3) seems ugly. On the other hand, (1) would clean up some existing code too!

The downsides of (1) are more static data and monomorphized code. The latter can be reduced if necessary by duplicating final kicks.

`legal-boards` will also need to handle Jstris 180 and TETRIO separately, because some Jstris placements will be illegal in TETRIO and vice versa.

**Comment by wirelyre** [comment]
After implementing (1), I backported it to the original pure SRS implementation.

With the new 180-compatible code:
- WASM blob size increased by ~1KiB (completely fine)
- Performance did not decrease
  - Based on unscientific stopwatch times running `gomen` on `*7` and `legal-boards`
  - It actually seems to be faster, but I'm not interested in measuring it precisely
  - The difference could be the shortcut for `O`

For now there are no downsides, so (1) is going into the codebase.

In the future it would be interesting to try completely runtime-chosen kick tables: Something like `(Vec<u8>, Vec<u64>)` rather than `(&'static [u8], &'static [u64])`.

**Comment by wirelyre** [comment]
Closed in d2440e83603a96262a80fd0bca14efb395bf6119.
